### ISSUE TITLE
Where the Gradient Descent takes place in the code?

### ISSUE THREAD

**Issue by Velmurugan1008** [william-redman/Koopman-Neural-Network-Training#1]
Hello, Sir! Good Evening.
I read your article "Optimizing Neural Networks via Koopman Operator Theory". I am interested to know about some topic in detail.
You said the NN trained based on the Gradient Descent algorithm at Equation (5) on page 3. May I know in which part of the MATLAB code the Gradient Descent is in action? Your explanations are beneficial for my understanding. Thank you for being so considerate.

**Comment by william-redman** [comment]
Thanks for your question! 

To provide some clarification, let me note that the MATLAB code provided does not perform gradient descent. Instead, it aims to learn a linear representation of gradient descent (Eq. 6 of our paper), from observations of weight updates. Once the Koopman operator is approximated ($\tilde{U}$ in our paper), training then simply involves multiplying $\tilde{U}$ to the weights. 

This means that, in order to use the MATLAB, you need to first have some weight evolution data. You can download an example data set here https://drive.google.com/file/d/1GTz0osiZttg1VAd0WQaTvaasOHq8btjL/view?usp=sharing. Having saved the example data to your computer, you can run HNN_master_example.m. This script calls the function NodeKoopmanTraining.m, which builds the Koopman operator and evolves the weights/biases forward. An error vs. weight/bias evolution plot (analogous to Fig. 1e) is produced. Various free parameters can be adjusted to get a feel for how Node Koopman training works.  

Feel free to let me know if you have more questions!

**Comment by william-redman** [comment]
I've closed this issue, but feel free to ask more questions on this thread.
